### PR TITLE
fix: null check error when using `withConverter` and returning null from `fromFirestore`

### LINKED ISSUE #7264: 🐛 [Firestore] withConverter doesn't work with nullable documents
## Bug report

**Describe the bug**
When using the ```withConverter``` function with a nullable type, if the ```fromFirestore``` returns a null (and it should be able to do that), the ```doc.data()``` throws an Exception.

### Steps to reproduce

1. Create a query using the ```withConverter``` method and a nullable class (i.e.: ```Data```). This query should have the following type: ```Query<Data?>```.
2. Get the query and map all documents to their datas: ```.map((doc) => doc.data())```
3. If any of ```fromFirestore``` calls returns null, it will throw an exception. Note: ```fromFirestore``` can return null if the map is empty or for any other reason (i.e.: the ).

### Expected behavior

Instead of throwing an exception, it should return the null value as it was intended by ```fromFirestore```. That way, instead of failing, the example above would result in ```Iterable<Data?>```.

### Sample project

```dart
import 'package:firebase/firestore.dart';

class Data {
  final String name;
  Data(this.name);
}

Stream<List<Data?>> get dataStream => FirebaseFirestore.instance
    .collection('data')
    .withConverter<Data?>(
      fromFirestore: (map, options) {
        final name = map['name'] as String?;
        if (name == null) return null;
        return Data(name);
      },
      toFirestore: (data, options) {
        assert(data != null);
        return {'name': data?.name};
      },
    )
    .snapshots()
    .map((snapshot) => snapshot.docs.map((doc) => doc.data()).toList());
```

---

## Additional context

I believe the problem is in the ```data()``` override in ```_WithConverterQueryDocumentSnapshot```:

```dart
class _WithConverterQueryDocumentSnapshot<T extends Object?>
    extends _WithConverterDocumentSnapshot<T>
    implements QueryDocumentSnapshot<T> {

/*...*/

  @override
  T data() => super.data()!;
  // This method results in an exception if ```T``` is nullable.
}
```

This is probably explained by the code below:

```dart
T force<T>(T? Function() data) {
  return data()!;
}
void main() {
  print(force<int?>(() => null));  // EXCEPTION
}
```

---

### Flutter doctor

Run `flutter doctor` and paste the output below:

<details><summary>Click To Expand</summary>

```
Doctor summary (to see all details, run flutter doctor -v):
[✓] Flutter (Channel stable, 2.5.1, on macOS 12.0.1 21A559 darwin-arm, locale en-BR)
[✓] Android toolchain - develop for Android devices (Android SDK version 30.0.3)
[✓] Xcode - develop for iOS and macOS
[✓] Chrome - develop for the web
[✓] Android Studio (version 2020.3)
[✓] VS Code (version 1.61.2)
[✓] Connected device (2 available)

• No issues found!
```

</details>

### PR DESCRIPTION
**Pull request by exaby73**
## Description

Returning null in `fromFirestore` causes a null check error to be thrown. This PR fixes that issue

## Related Issues

Fixes #7264

## Checklist

Before you create this PR confirm that it meets all requirements listed below by checking the relevant checkboxes (`[x]`).
This will ensure a smooth and quick review process. Updating the `pubspec.yaml` and changelogs is not required.

- [ ] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [ ] My PR includes unit or integration tests for *all* changed/updated/fixed behaviors (See [Contributor Guide]).
- [ ] All existing and new tests are passing.
- [ ] I updated/added relevant documentation (doc comments with `///`).
- [ ] The analyzer (`melos run analyze`) does not report any problems on my PR.
- [ ] I read and followed the [Flutter Style Guide].
- [ ] I signed the [CLA].
- [ ] I am willing to follow-up on review comments in a timely manner.

## Breaking Change

Does your PR require plugin users to manually update their apps to accommodate your change?

- [ ] Yes, this is a breaking change.
- [x] No, this is *not* a breaking change.

[issue database]: https://github.com/flutter/flutter/issues
[Contributor Guide]: https://github.com/firebase/flutterfire/blob/master/CONTRIBUTING.md
[Flutter Style Guide]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo
[pub versioning philosophy]: https://dart.dev/tools/pub/versioning
[CLA]: https://cla.developers.google.com/

### CODE REVIEW

File: packages/cloud_firestore/cloud_firestore/example/integration_test/collection_reference_e2e.dart [hunk 1]
```diff
@@ -169,6 +169,32 @@ void runCollectionReferenceTests() {
         },
         timeout: const Timeout.factor(3),
       );
+
+      testWidgets(
+        '#8082 - returning null from `fromFirestore` should not throw a null check error',
```
Comment (Lyokone) [inline packages/cloud_firestore/cloud_firestore/example/integration_test/collection_reference_e2e.dart:174]: We are not putting issues number in the test description (we already have this information with git history)

File: packages/cloud_firestore/cloud_firestore/example/ios/Runner/GoogleService-Info.plist [hunk 2]
```diff
@@ -35,4 +35,4 @@
 	<key>DATABASE_URL</key>
 	<string>https://flutterfire-e2e-tests-default-rtdb.europe-west1.firebasedatabase.app</string>
 </dict>
-</plist>
\ No newline at end of file
+</plist>
```
Comment (russellwheatley) [inline packages/cloud_firestore/cloud_firestore/example/ios/Runner/GoogleService-Info.plist:38]: can also remove this change.

**Review by russellwheatley — CHANGES_REQUESTED** [review]

File: packages/cloud_firestore/cloud_firestore/example/ios/Runner/GoogleService-Info.plist (hunk 2 above)
Comment (exaby73) [reply packages/cloud_firestore/cloud_firestore/example/ios/Runner/GoogleService-Info.plist:38]: Honestly, I am not sure what changed here. I literally tried copying the code from master and pasting it. I'm confused

File: packages/cloud_firestore/cloud_firestore/example/integration_test/collection_reference_e2e.dart (hunk 1 above)
Comment (exaby73) [reply packages/cloud_firestore/cloud_firestore/example/integration_test/collection_reference_e2e.dart]: Done

File: packages/cloud_firestore/cloud_firestore/example/ios/Runner/GoogleService-Info.plist (hunk 2 above)
Comment (exaby73) [reply packages/cloud_firestore/cloud_firestore/example/ios/Runner/GoogleService-Info.plist]: Done :D

**Comment by cedvdb** [comment]
`R Function(DocumentSnapshot<Map<String, dynamic>>, SnapshotOptions?) fromFirestore`

should be

`R? Function(DocumentSnapshot<Map<String, dynamic>>, SnapshotOptions?) fromFirestore`

**Comment by cedvdb** [comment]
This is not fixed.

```
typedef FromFirestore<T> = T Function(
  DocumentSnapshot<Map<String, dynamic>> snapshot,
  SnapshotOptions? options,
);
```
Should be

```
typedef FromFirestore<T> = T? Function(
  DocumentSnapshot<Map<String, dynamic>> snapshot,
  SnapshotOptions? options,
);
```

The ToFirestore should not be nullable on the other hand.

**Pull request merged**